File: JavaPractice/src/MaxHeapArrayCanReturnTopLargestElement.java
public class MaxHeapArrayCanReturnTopLargestElement {

	private static MaxHeapArrayCanReturnTopLargestElement maxHeap;
	private int[] heap;
	private int capacity;
	private int position;
	
	public MaxHeapArrayCanReturnTopLargestElement(int cap) {
		position = 1;
		capacity = cap;
		heap = new int[capacity];
	}
	
	public void insert(int val) {
		if(position == capacity) {
			
		}
		else {
			heap[position++] = val;
			
			int child = position - 1;
			int parent = child / 2;
			
			while(heap[parent] < heap[child] && parent > 0) {
				int temp = heap[parent];
				heap[parent] = heap[child];
				heap[child] = temp;
				
				child = parent;
				parent = child/2;
			}
			
		}
	}
	
	public void print() {
		for(int i = 1; i < position; i++) {
			System.out.print(heap[i] + " ");
		}
	}
	
	public static void main(String[] args) {
		maxHeap = new MaxHeapArrayCanReturnTopLargestElement(10);
		maxHeap.insert(12);
		maxHeap.insert(7);
		maxHeap.insert(6);
		maxHeap.insert(10);
		maxHeap.insert(8);
		maxHeap.insert(20);
		maxHeap.print();
	}
}

File: PythonPractice/src/MaxHeapArrayCanReturnTopLargestElement.py
class MaxHeap:
    def __init__(self, capacity):
        if not isinstance(capacity, int) or capacity <= 0:
            raise ValueError("Capacity must be a positive integer")
        self.capacity = capacity
        self.heap = [0] * (capacity + 1)  # Index 0 is unused
        self.position = 1

    def insert(self, val):
        if not isinstance(val, int):
            raise ValueError("Value to insert must be an integer")

        if self.position >= self.capacity:
            raise OverflowError("Heap is full")

        self.heap[self.position] = val
        child = self.position
        parent = child // 2

        while parent > 0 and self.heap[parent] < self.heap[child]:
            # Swap parent and child
            self.heap[parent], self.heap[child] = self.heap[child], self.heap[parent]

            child = parent
            parent = child // 2

        self.position += 1

    def print_heap(self):
        # Print the heap ignoring index 0
        print(" ".join(map(str, self.heap[1:self.position])))

if __name__ == "__main__":
    max_heap = MaxHeap(10)
    max_heap.insert(12)
    max_heap.insert(7)
    max_heap.insert(6)
    max_heap.insert(10)
    max_heap.insert(8)
    max_heap.insert(20)
    max_heap.print_heap()
